cmd/shortener/main.go

```diff
@@ -28,25 +44,69 @@ func run() error {
 		return err
 	}
 
-	storage, err := getStorage(cfg)
+	storage, err := getStorage(ctx, cfg)
 	if err != nil {
 		return err
 	}
-
+	sessionStorage := appStorage.NewSessionStorage()
 	shortURLService := url.NewShortURLService(storage)
-	logger.LogSugar.Infof("Running server on - %s", cfg.ServerURL)
 	stop := make(chan struct{})
-	return http.ListenAndServe(cfg.ServerURL, handlers.AppRoutes(shortURLService, stop))
+	worker := workers.NewWorker(storage, stop)
+	routes := handlers.NewRoutes(shortURLService, storage, sessionStorage, worker).Init()
+
+	if cfg.PprofEnabled {
+		routes.Mount("/debug", middleware.Profiler())
+	}
+
+	httpServer := http.Server{
+		Addr:    cfg.ServerURL,
+		Handler: routes,
+		BaseContext: func(net.Listener) context.Context {
+			return ctx
+		},
+	}
+	go func() {
+		<-ctx.Done()
+		// Отправка сигнала о завершении в канал воркерам
+		stop <- struct{}{}
+		logger.LogSugar.Info("Получин сигнал. Останавливаю сервер...")
+
+		shutdownCtx, cancel := context.WithTimeout(ctx, 15*time.Second)
+		defer cancel()
+		err = httpServer.Shutdown(shutdownCtx)
+		if err != nil {
+			logger.LogSugar.Error(err)
+		}
+	}()
+
+	logger.LogSugar.Infof("Running server on - %s", cfg.ServerURL)
+	err = httpServer.ListenAndServe()
+	if err != nil && !errors.Is(err, http.ErrServerClosed) {
+		return err
+	}
+	if errors.Is(err, http.ErrServerClosed) {
+		logger.LogSugar.Info("Сервер остановлен")
+	}
+
+	return nil
```

Comment:
Мне кажется тут можно упростить:
```suggestion
	if err != nil {
	        if errors.Is(err, http.ErrServerClosed) {
		        logger.LogSugar.Info("Сервер остановлен")
		        return nil
	        }
		return err
	}

```